### PR TITLE
Introduce plotly directive.

### PR DESCRIPTION
**Pull request by harupy**
Propose a `plotly` directive which works the same as matplotlib's `plot` directive.

## Motivation

Pros:

We can:
- remove `plot_xxx.py` and `plot_xxx.html` under `docs/source`.
- prevent mismatch between example code and its output.
- automatically update plotly figures on the doc.

Cons:
- The doc build time would increase.
- ~We have to maintain `plot_directive.py`.~

## Description of the changes

- ~Add `plotly_directive.py`.~
- Add `sphinx-plotly-directive` to `extras_require: document`.
- Apply the directive to `optuna/visualization/_edf.py` to verify it works properly.

### CODE REVIEW

**Comment by harupy** [comment]
Result:

https://61319-122299416-gh.circle-artifacts.com/0/docs/build/html/reference/visualization/generated/optuna.visualization.plot_edf.html#optuna.visualization.plot_edf

File: docs/source/plotly_directive.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+def run_code(code):
+    namespace = {}
+    variable_name = "fig"
+
+    try:
+        exec(assign_last_line_into_variable(code, variable_name), namespace)
+    except (Exception, SystemExit) as err:
+        raise PlotError(traceback.format_exc()) from err
+
+    return namespace[variable_name]
```
Comment (harupy) [inline docs/source/plotly_directive.py:109]: ## How to extract plotly figure from example code block:

Let's say we have code like this:
```
...
func_that_returns_plotly_figure()
```

`assign_last_line_into_variable` injects `fig = ` in the last line:

```
...
fig = func_that_returns_plotly_figure() 
```

Running this code generates a variable named `fig` which stores a plotly figure object.

File: docs/source/plotly_directive.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    go.Figure(data=go.Scatter(x=t, y=y, mode='markers'))
+```
+
+Requirements:
+1. docstring contains a single code block.
+2. the code block ends with an expression that evaluates to a plotly figure.
+
+Based on:
+https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/sphinxext/plot_directive.py
```
Comment (harupy) [inline docs/source/plotly_directive.py:20]: Matplotlib's plot_directive is much more complicated because it supports multile code blocks and figures.

**Comment by harupy** [comment]
I created a package for `plotly` directive:
https://github.com/harupy/sphinx-plotly-directive

**Review by keisuke-umezawa — COMMENTED** [review]
@harupy If we use this directive, `doctest` will not run it if my understanding is correct. Is it okay?
(In my opinion, it should be okay, because the code in `plotly` directive is run during generating documentation.)

**Review by keisuke-umezawa — APPROVED** [review]

**Comment by harupy** [comment]
@keisuke-umezawa Thanks for the review.

> If we use this directive, doctest will not run it if my understanding is correct.

Yes, your understanding is correct. doctest ignores code blocks annotated with `plotly` (and `plot`) directives.

**Comment by harupy** [comment]
If the proposed change is acceptable, I'll apply the `plotly` directive to the rest of visualization functions.

**Comment by toshihikoyanase** [comment]
Thank you for your great PR!
The change looks good to me, but I have a concern about the license of `sphinx-plotly-directive`. I think [the repository](https://github.com/harupy/sphinx-plotly-directive) does not contain the `LICENSE` file, so could add it?

File: docs/source/conf.py [hunk 3]
```diff
@@ -193,3 +194,9 @@
 plot_formats = [("png", 90)]
 plot_html_show_formats = False
 plot_html_show_source_link = False
+
+# sphinx plotly directive
+plotly_include_source = True
+plotly_formats = ["html"]
```
Comment (toshihikoyanase) [inline docs/source/conf.py:200]: [Question] I think this setting tells `plotly` to generate figures as HTML + JS, and I don't think it requires [`Orca`](https://github.com/plotly/orca) or [`Kaleido`](https://github.com/plotly/Kaleido). Is that correct?

Comment (harupy) [reply docs/source/conf.py:200]: Yes, correct :)

Comment (toshihikoyanase) [reply docs/source/conf.py:200]: Thanks for your swift response!

**Comment by harupy** [comment]
The contour plot looks strange:

https://62001-122299416-gh.circle-artifacts.com/0/docs/build/html/reference/visualization/generated/optuna-visualization-plot_contour-1.html

**Comment by HideakiImamura** [comment]
@harupy Thanks you for the PR and investigation. I don't know why the problem of the strange contour plot occurs, but I confirmed that the example code ([here](https://optuna.readthedocs.io/en/latest/reference/visualization/generated/optuna.visualization.plot_contour.html)) master generates the following plot with the current master branch. Did you use the same code?

スクリーンショット 2020-10-27 9 30 06

**Comment by harupy** [comment]
@HideakiImamura 
The current master just renders `docs/source/plotly_figures/plot_contour.html` and doesn't use the example code at all.

#1819 was merged on Sep 12, but `plot_contour.html` hasn't been updated since August 30 ([link](https://github.com/optuna/optuna/blob/master/docs/source/plotly_figures/plot_contour.html)).

**Comment by HideakiImamura** [comment]
Sorry for the lack of explanation. I am aware that the image shown in the documentation has not been updated. The image I just attached is the one I got by running the example code in local with the current master branch.

**Comment by harupy** [comment]
Strange. Can you run this code and check what it outputs?

```python
import optuna


def objective(trial):
    x = trial.suggest_uniform("x", -100, 100)
    y = trial.suggest_categorical("y", [-1, 0, 1])
    return x ** 2 + y


study = optuna.create_study()
study.optimize(objective, n_trials=30)

print(optuna.visualization.plot_contour(study, params=["x", "y"]))

```

**Comment by HideakiImamura** [comment]
In my environment, the output looks like this.
```python
Figure({
    'data': [{'colorbar': {'title': {'text': 'Objective Value'}},
              'colorscale': [[0, 'rgb(220,220,220)'], [0.30000000000000004,
                             'rgb(106,137,247)'], [0.4, 'rgb(90,120,245)'], [0.5,
                             'rgb(70,100,245)'], [0.65, 'rgb(40,60,190)'], [1,
                             'rgb(5,10,172)']],
              'connectgaps': True,
              'contours': {'coloring': 'heatmap'},
              'hoverinfo': 'none',
              'line': {'smoothing': 1.3},
              'type': 'contour',
              'x': [-100.62804209298965, -91.91372303891319, -65.33686738799064,
                    -60.68923158205577, -47.46322887329877, -41.31464469477148,
                    -31.605102435181198, -24.659250870156505, -24.073711734782858,
                    -18.19159020334969, -14.228753212041429, -9.285265073503444,
                    -6.570538307972498, -5.58755386027822, -4.110352081422276,
                    -2.6062176084841635, -0.8715970258380716, 1.9181247121672307,
                    6.133031579934413, 8.141212630960725, 12.131100674217592,
                    21.893792812880434, 22.62421834339918, 24.520275630325013,
                    39.295447059682544, 42.60662213389958, 43.53512360492172,
                    48.16946014693917, 68.36170759989018, 81.86664235066142,
                    82.37265804261588, 91.08697709669234],
              'y': [-1, 0, 1],
              'z': [[nan, 8447.132482874042, 4267.906240075875, nan, nan,
                    1705.8998662552092, nan, nan, nan, 329.93395412660846, nan,
                    nan, nan, nan, nan, nan, nan, 2.6792024114266217, nan,
                    65.27934310251445, 146.16360356800251, nan, 510.85525564979986,
                    nan, nan, nan, nan, 2319.2968908475614, nan,
                    6701.1471297711105, nan, nan], [nan, nan, nan,
                    3683.182830020395, nan, nan, 998.8824999382965,
                    608.0786534773143, nan, nan, 202.4574179691793,
                    86.21614748522292, 43.1719736565341, 31.22075814151004,
                    16.894994233252437, 6.792370222772912, 0.7596813754497721, nan,
                    37.61407636047281, nan, nan, 479.33816373333497, nan,
                    601.2439169871108, nan, 1815.3242496609014, nan, nan,
                    4673.323065972883, nan, 6785.254793005731, nan], [nan, nan,
                    nan, nan, 2253.7580950791416, nan, nan, nan, 580.5435966894219,
                    nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
                    nan, nan, nan, 1545.1321596203136, nan, 1896.3069872958124,
                    nan, nan, nan, nan, nan]]},
             {'marker': {'color': 'black', 'line': {'color': 'Grey', 'width': 0.5}},
              'mode': 'markers',
              'showlegend': False,
              'type': 'scatter',
              'x': [-6.570538307972498, 12.131100674217592, 22.62421834339918,
                    39.295447059682544, -91.91372303891319, 82.37265804261588,
                    48.16946014693917, 42.60662213389958, 43.53512360492172,
                    -47.46322887329877, -31.605102435181198, -14.228753212041429,
                    1.9181247121672307, -65.33686738799064, -5.58755386027822,
                    81.86664235066142, -24.659250870156505, -60.68923158205577,
                    8.141212630960725, -41.31464469477148, 24.520275630325013,
                    -9.285265073503444, -4.110352081422276, -0.8715970258380716,
                    21.893792812880434, -24.073711734782858, 68.36170759989018,
                    6.133031579934413, -18.19159020334969, -2.6062176084841635],
              'y': [0, -1, -1, 1, -1, 0, -1, 0, 1, 1, 0, 0, -1, -1, 0, -1, 0, 0,
                    -1, -1, 0, 0, 0, 0, 0, 1, 0, 0, -1, 0]}],
    'layout': {'template': '...',
               'title': {'text': 'Contour Plot'},
               'xaxis': {'range': [-100.62804209298965, 91.08697709669234], 'title': {'text': 'x'}},
               'yaxis': {'range': ['-1', '1'], 'title': {'text': 'y'}, 'type': 'category'}}
})
```

**Comment by HideakiImamura** [comment]
I run the following code and got the following image.
```python
import optuna


def objective(trial):
    x = trial.suggest_uniform("x", -100, 100)
    y = trial.suggest_categorical("y", [-1, 0, 1])
    return x ** 2 + y


study = optuna.create_study()
study.optimize(objective, n_trials=30)

fig = optuna.visualization.plot_contour(study, params=["x", "y"])
print(fig)
fig.show()
```
スクリーンショット 2020-10-27 12 02 48

**Comment by harupy** [comment]
- This issue occurs in plotly.py 4.12.0.
- plotly.py 4.12.0 uses plotly.js 1.57.1.
- plotly.js 1.57. contains https://github.com/plotly/plotly.js/pull/5211 which seems to have changed how to compute a range for a categorical parameter.

**Comment by HideakiImamura** [comment]
Thanks for the investigation. In my environment, plotly is the version 4.9.0. I confirmed that the strange contour plot is generated with the latest version of plotly. But, as you said, this bug is not related to the change of this PR. I will create an issue for this problem. Thanks again for your great contribution!

**Comment by harupy** [comment]
I'll submit two follow-up PRs once this one has been merged:

1. Remove `plot_xxx.py` and `plot_xxx.html` under `docs`.
2. Make the examples deterministic.

**Comment by harupy** [comment]
@toshihikoyanase @HideakiImamura Could you reivew this PR again?

**Review by toshihikoyanase — APPROVED** [review]
@harupy @HideakiImamura  Thank you for the investigation of #1967. I confirmed that the unexpected contour plot is not the problem of `plotly` directive. When I applied #1983, the contour plot was rendered as expected.

So, the change looks good to me and I think we can merge this PR and work on #1967 in follow-up PRs (including #1983). What do you think?

image

**Comment by harupy** [comment]
Sounds good :)

**Pull request merged**